File: 1-easy/problems/3838-weighted-word-mapping/solution.cpp
// Time:  O(n * l)
// Space: O(1)

// string
class Solution {
public:
    string mapWordWeights(vector<string>& words, vector<int>& weights) {
        string result;
        result.reserve(size(words));
        for (const auto& w: words) {
            int i = 0;
            for (const auto& x : w) {
                i = (i + weights[x - 'a']) % 26;
            }
            result.push_back('z' - i);
        }
        return result;
    }
};

File: 1-easy/problems/3838-weighted-word-mapping/solution.py
from typing import List, Optional, Dict, Set, Tuple

class Solution:
    def mapWordWeights(self, words, weights):
        """
        :type words: List[str]
        :type weights: List[int]
        :rtype: str
        """
        result = []
        for w in words:
            i = 0
            for x in w:
                i = (i+weights[ord(x)-ord('a')])%26
            result.append(chr(ord('z')-i))
        return "".join(result)


# string
class Solution2(object):
    def mapWordWeights(self, words, weights):
        """
        :type words: List[str]
        :type weights: List[int]
        :rtype: str
        """
        return "".join(chr(ord('z')-reduce(lambda accu, x: (accu+x)%26, map(lambda x: weights[ord(x)-ord('a')], w), 0)) for w in words)
